Add indices

jakirkham opened the pull request:
Fixes https://github.com/dask/dask/issues/2267

Adds an equivalent function to [`numpy.indices`]( https://docs.scipy.org/doc/numpy/reference/generated/numpy.indices.html ).

mrocklin commented on dask/array/creation.py:
```diff
@@ -135,3 +137,42 @@ def arange(*args, **kwargs):
+            r = r[s]
+
+            for j in chain(range(i), range(i + 1, len(dimensions))):
+                r = r.repeat(dimensions[j], axis=j).rechunk({j: chunks[j + 1]})
```
The repeated rechunking here might be expensive in terms of both graph creation and numpy copies.

Also a slight preference to use `itertools.chain` rather than `chain`

mrocklin replied:
Hrm, is it possible to get by with arange and slicing (as you've done above), and multiplying by `da.ones(full_shape, chunks=full_chunks)`?  This might also handle the rechunking automatically.  This multiplication is probably more expensive on the workers than is repeat, but I suspect that this isn't that significant and that it's cheaper when constructing the graph

jakirkham replied:
Was initially worried as I saw some chunking by 1 in earlier code, but it seems `repeat` here actually results in chunks equal to the number of repeats. Once everything is stacked, it does appear to get the right chunking for everything except the first dimension. However, `stack` does need to be rechunked on the first dimension as that is chunked by 1.

jakirkham replied:
Also should add it does seem to go a little faster by dropping the chunking after `repeat` based on some simple testing.

jakirkham commented on dask/array/creation.py:
```diff
@@ -135,3 +137,42 @@ def arange(*args, **kwargs):
+            grid.append(r)
+
+    if grid:
+        grid = stack(grid).rechunk({0: chunks[0]})
```
I suppose we could actually have chunks just be 1 here. Then the user would only specify chunks for the rest of the array. That way there is no need to rechunk. Thoughts?

jakirkham replied:
Counterpoint would be this could be confusing to the user to not specify chunks of the same dimensionality of the expected shape.

mrocklin replied:
So we have k arrays of the same shape and we want to stack them along a new axis.  The question is if we should chunk those arrays together or leave them as is?  If so then leaving them as is seems fine to me.

jakirkham replied:
Agreed. That simplifies some code as well. Have pushed a change.

jakirkham commented:
Have addressed your comments @mrocklin. Also have raised a question of my own above.

This seems to be passing except for one failure due to the `test_interrupt` issue ( https://github.com/dask/dask/issues/2192 ).

jakirkham commented on dask/array/creation.py:
```diff
@@ -135,3 +137,40 @@ def arange(*args, **kwargs):
     return Array(dsk, name, chunks, dtype=dtype)
+
+
+@wraps(np.indices)
```
I basically got this from copy-n-pasting something from `dask.array.core`. Is there any value in having this for Dask beyond getting the docstring? If not, was planning to replace the docstring given our parameters differ.

mrocklin replied:
Replacing with a custom docstring is preferred if you're willing.

jakirkham replied:
Cool, already have something ready to go. 😉

jakirkham commented:
Well I can't think of anything else needed on my end. Anything else you'd like to see @mrocklin or is this good to go?

jakirkham commented:
Hmm...not sure why the AppVeyor tests mysteriously stopped, but it seems unrelated Unfortunately no more information is provided by the log.

ref: https://ci.appveyor.com/project/daskdev/dask/build/1.0.1180#L467

jakirkham commented:
Repushed and it seems everything is green.

mrocklin commented:
Thanks @jakirkham !

jakirkham commented:
Thanks @mrocklin. Just realized I forgot to add this to the docs. Fixing that with PR ( https://github.com/dask/dask/pull/2275 ).
